File: typescript/src/Portfolio.ts
import {Asset, MeasurableValue, PricelessValue} from "./Asset";
import * as fs from "fs";

export class Portfolio {
  private readonly _portfolioCsvPath: string;

  constructor(portfolioCsvPath: string) {
    this._portfolioCsvPath = portfolioCsvPath;
  }

  computePortfolioValue(): void {
    const now = new Date();
    const readText = fs.readFileSync(this._portfolioCsvPath, {encoding: 'utf8'});
    const lines = readText.split(/\r?\n/);
    let portfolioValue = new MeasurableValue(0);

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const columns = line.split(",");

      const [year, month, day] = columns[1].split("/");
      if ([year, month, day].some((value) => value === undefined)) {
        throw new Error("wrong date");
      }
      const date = new Date(Number(year), Number(month) - 1, Number(day));

      const asset = new Asset(columns[0], date,
        columns[0] == "Unicorn" ? new PricelessValue() : new MeasurableValue(Number(columns[2])));

      if (Math.floor((asset.getDate().getTime() - now.getTime()) / (1000 * 60 * 60 * 24)) < 0) {
        if (asset.getDescription() != "French Wine") {
          if (asset.getDescription() != "Lottery Prediction") {
            if (asset.getValue().get() > 0) {
              if (asset.getDescription() != "Unicorn") {
                asset.setValue(new MeasurableValue(asset.getValue().get() - 20));
              } else {
                console.log(
                  "Portfolio is priceless because it got a unicorn on " +
                  asset.getDate() + "!!!!!");
                return;
              }
            }
          } else {
            asset.setValue(new MeasurableValue(asset.getValue().get() - asset.getValue().get()));
          }
        } else {
          if (asset.getValue().get() < 200) {
            asset.setValue(new MeasurableValue(asset.getValue().get() + 20));
          }
        }
      } else {
        if (asset.getDescription() != "French Wine" && asset.getDescription() != "Lottery Prediction") {
          if (asset.getValue().get() > 0.0) {
            if (asset.getDescription() != "Unicorn") {
              asset.setValue(new MeasurableValue(asset.getValue().get() - 10));
            } else {
              console.log("Portfolio is priceless because it got a unicorn on " + asset.getDate() + "!!!!!");
              return;
            }
          } else {
            if (asset.getDescription() == "Unicorn") {
              console.log("Portfolio is priceless because it got a unicorn on " + asset.getDate() + "!!!!!");
              return;
            }
          }
        } else {
          if (asset.getDescription() == "Lottery Prediction") {
            if (asset.getValue().get() < 800) {
              asset.setValue(new MeasurableValue(asset.getValue().get() + 5));

              if (Math.floor((asset.getDate().getTime() - now.getTime()) / (1000 * 60 * 60 * 24)) < 11) {
                if (asset.getValue().get() < 800) {
                  asset.setValue(new MeasurableValue(asset.getValue().get() + 20));
                }
              }

              if (Math.floor((asset.getDate().getTime() - now.getTime()) / (1000 * 60 * 60 * 24)) < 6) {
                if (asset.getValue().get() < 800) {
                  asset.setValue(new MeasurableValue(asset.getValue().get() + 100));
                }
              }
            }
          } else {
            if (asset.getValue().get() < 200) {
              asset.setValue(new MeasurableValue(asset.getValue().get() + 10));
            }
          }
        }
      }
      portfolioValue = new MeasurableValue(portfolioValue.get() + asset.getValue().get());
    }
    console.log(portfolioValue);
  }
}

File: c_sharp/Portfolio/Portfolio.cs
using static System.Globalization.CultureInfo;

namespace Portfolio;

public class Portfolio
{
    private readonly string _portfolioCsvPath;

    public Portfolio(string portfolioCsvPath)
    {
        _portfolioCsvPath = portfolioCsvPath;
    }

    public void ComputePortfolioValue()
    {
        var now = DateTime.Now;
        var readText = File.ReadAllText(_portfolioCsvPath);
        var lines = readText.Split(Environment.NewLine);
        var portfolioValue = new MeasurableValue(0);

        foreach (var line in lines)
        {
            var columns = line.Split(",");
            var asset = new Asset(columns[0],
                DateTime.Parse(columns[1], CurrentCulture),
                columns[0] == "Unicorn" ? new PricelessValue() : new MeasurableValue(int.Parse(columns[2])));

            if (asset.Date.Subtract(now).TotalDays < 0)
            {
                if (asset.Description != "French Wine")
                {
                    if (asset.Description != "Lottery Prediction")
                    {
                        if (asset.Value.Get() > 0)
                        {
                            if (asset.Description != "Unicorn")
                            {
                                asset.Value = new MeasurableValue(asset.Value.Get() - 20);
                            }
                            else
                            {
                                Console.WriteLine(
                                    "Portfolio is priceless because it got a unicorn on " +
                                    asset.Date.ToString(CurrentCulture) + "!!!!!");
                                return;
                            }
                        }
                    }
                    else
                    {
                        asset.Value = new MeasurableValue(asset.Value.Get() - asset.Value.Get());
                    }
                }
                else
                {
                    if (asset.Value.Get() < 200) asset.Value = new MeasurableValue(asset.Value.Get() + 20);
                }
            }
            else
            {
                if (asset.Description != "French Wine" && asset.Description != "Lottery Prediction")
                {
                    if (asset.Value.Get() > 0.0)
                    {
                        if (asset.Description != "Unicorn")
                        {
                            asset.Value = new MeasurableValue(asset.Value.Get() - 10);
                        }
                        else
                        {
                            Console.WriteLine(
                                "Portfolio is priceless because it got a unicorn on " +
                                asset.Date.ToString(CurrentCulture) + "!!!!!");
                            return;
                        }
                    }
                    else
                    {
                        if (asset.Description == "Unicorn")
                        {
                            Console.WriteLine(
                                "Portfolio is priceless because it got a unicorn on " +
                                asset.Date.ToString(CurrentCulture) + "!!!!!");
                            return;
                        }
                    }
                }
                else
                {
                    if (asset.Description == "Lottery Prediction")
                    {
                        if (asset.Value.Get() < 800)
                        {
                            asset.Value = new MeasurableValue(asset.Value.Get() + 5);

                            if (asset.Date.Subtract(now).TotalDays < 11)
                                if (asset.Value.Get() < 800)
                                    asset.Value = new MeasurableValue(asset.Value.Get() + 20);

                            if (asset.Date.Subtract(now).TotalDays < 6)
                                if (asset.Value.Get() < 800)
                                    asset.Value = new MeasurableValue(asset.Value.Get() + 100);
                        }
                    }
                    else
                    {
                        if (asset.Value.Get() < 200) asset.Value = new MeasurableValue(asset.Value.Get() + 10);
                    }
                }
            }

            portfolioValue = new MeasurableValue(portfolioValue.Get() + asset.Value.Get());
        }

        Console.WriteLine(portfolioValue);
    }
}